### PR TITLE
core: restructure types for direct import and publishing

### PR DESCRIPTION
**Pull request by adamraine**
Proposal that helps us expose our types #1773 

- Types are generated the same way as before, but now they can be `rsync`ed into the normal directory structure with `yarn build-types`. This is done in prepack so should be in the release automatically.
- `*.d.ts` that just modify the global namespace or module declarations are moved to the `types/env` directory (i.e. types only we use).
- `*.js` will need to do `import * as LH from '../types/lh.js'` in order for their generated type declarations to be valid. (see `core/index.js` for example)
- Remove usages of global `LH` from core `*.d.ts` files
- Requires the `.js` extension when importing `*.d.ts` files in core
- Closes https://github.com/GoogleChrome/lighthouse/issues/14654

### CODE REVIEW

File: types/env/global.d.ts [hunk 1]
```diff
@@ -0,0 +1,3 @@
+export * from '../lh';
+
+export as namespace LH;
```
Comment (adamraine) [inline types/env/global.d.ts:3]: For this to work, we can't expose `LH` globally. This is a temporary measure if we decide to split into multiple PRs.

File: types/externs.ts [hunk 2]
```diff
@@ -4,7 +4,7 @@
  * Unless required by applicable law or agreed to in writing, software distributed under the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the License for the specific language governing permissions and limitations under the License.
  */
 
-import {Artifacts} from './artifacts';
```
Comment (adamraine) [inline types/externs.ts:7]: This file is kinda weird. It houses some necessary type definitions, but also modifies the global namespace. I propose we remove this file, move the type definitions to a `*.ts` file and the global namespace stuff to `node.d.ts` or something.

File: types/user-flow.ts [hunk 3]
```diff
@@ -14,6 +14,6 @@ declare module UserFlow {
   }
 }
 
-type UserFlow = typeof UserFlow_;
+type UserFlow = InstanceType<typeof UserFlow_>;
```
Comment (adamraine) [inline types/user-flow.ts:17]: This was a drive-by fix I found while testing this

Comment (adamraine) [reply types/user-flow.ts:17]: Actually I'll just move this to a separate PR

**Comment by brendankenny** [comment]
I had assumed we would solve #1773 with a limited set of types (our public interface), but since we're approaching year 6 for that issue and still no public API types, we should absolutely consider this approach :)

`d.ts` generation for all our js files already works great (that's how the incremental build and inter-project references work at all), so why not just use them?

A few thoughts:
- separating Lighthouse types from global augmentation (and typing npm dependencies) generally seems like a really good idea, though there will likely be some questionable overlap depending on what someone imports from Lighthouse since we're exposing _everything_
- moving types to (apparent) source code `ts` files seems confusing to me. We already have `d.ts` files for those types and don't need tsc to generate them, just to move them to the types directory. We could do that ourselves while keeping our type declarations as ambient `d.ts` files
- alternatively: we can leave them right in place and use the `d.ts` files next to their source `js` files approach and it should all just work. We can have tsc handle that, just need to remove the [outDir line in `tsconfig-base.json`](https://github.com/GoogleChrome/lighthouse/blob/78b93aacacb12ae10f14049c5a16bc48a431f5a6/tsconfig-base.json#L11) so declarations are written in place (or invoke `tsc` with `--outDir .`, though there are complications with `--build`).
- the `import * as LH from '../types/lh.js'` hack is a really clever way around the lack of jsdoc * imports, but is also pretty gross. Not sure I love it yet :) I wonder if we can get the ts team to finally implement a fix so at least we'd have a limited lifespan of those import LHs around. There's probably also a bunch of files that only need like one thing on `LH` and could do the direct typedef import instead of the giant namespace.
- you might need to flesh out more of this implementation to flush out some other issues with exporting everything. e.g. I'm pretty sure several of our devDependencies will need to become dependencies without some reworking. `puppeteer` is a big one with the current puppeteer.d.ts. I'm also not completely sure how it'll work with a `types` field in package.json since our `main` isn't currently in the project root. Maybe `d.ts` files will have to be side by side?

**Comment by adamraine** [comment]
> alternatively: we can leave them right in place and use the d.ts files next to their source js files approach and it should all just work. We can have tsc handle that, just need to remove the [outDir line in tsconfig-base.json](https://github.com/GoogleChrome/lighthouse/blob/78b93aacacb12ae10f14049c5a16bc48a431f5a6/tsconfig-base.json#L11) so declarations are written in place (or invoke tsc with --outDir ., though there are complications with --build).

This idea sounds fine to me. Personally I like the filesystem organization where `d.ts` files are in a separate directory but it wouldn't be that hard to deal with `d.ts` alongside every `.js` file.

> moving types to (apparent) source code ts files seems confusing to me. We already have d.ts files for those types and don't need tsc to generate them, just to move them to the types directory. We could do that ourselves while keeping our type declarations as ambient d.ts files

I think this depends on where we decide to put the generated `d.ts` files. If they are next to the `.js` files then we don't need to create a second copy because the path to our handwritten `d.ts` files will be the same for the generated `d.ts` and source `.js` files.

If the generated `d.ts` files are going in a separate directory, I want to avoid having a custom build step that moves our handwritten `d.ts` files to the new directory. Letting `tsc` handle this would be better IMO.

> the import * as LH from '../types/lh.js' hack is a really clever way around the lack of jsdoc * imports, but is also pretty gross. Not sure I love it yet :) I wonder if we can get the ts team to finally implement a fix so at least we'd have a limited lifespan of those import LHs around. There's probably also a bunch of files that only need like one thing on LH and could do the direct typedef import instead of the giant namespace.

I agree it's not ideal. However the alternative is to painstakingly go through every file and import the required types manually. Adding an `import * as LH from '...'` to every js file is something we could automate and do once.

> you might need to flesh out more of this implementation to flush out some other issues with exporting everything. e.g. I'm pretty sure several of our devDependencies will need to become dependencies without some reworking. puppeteer is a big one with the current puppeteer.d.ts. I'm also not completely sure how it'll work with a types field in package.json since our main isn't currently in the project root. Maybe d.ts files will have to be side by side?

This is a good point. I intended this to be a follow up but it's not a bad idea to look into it now.

**Comment by brendankenny** [comment]
> This idea sounds fine to me. Personally I like the filesystem organization where `d.ts` files are in a separate directory but it wouldn't be that hard to deal with `d.ts` alongside every `.js` file.

Just in case we're speaking past each other, I meant as a pack/publish step. It's not useful to have them around otherwise.

> If the generated `d.ts` files are going in a separate directory, I want to avoid having a custom build step that moves our handwritten `d.ts` files to the new directory. Letting `tsc` handle this would be better IMO.

Mixing ts and js files is pretty bad when the ts files aren't intended as source files, IMO, and makes the codebase less self evident. We're using `d.ts` files as the ambient declarations they already are, and it's good to have things enforced like no implementation allowed inside.

But I'm like 80% sure the question is moot because having more than one importable path with an explicit `types` is tricky and we'll probably have to do the in-place `d.ts` files anyways.

> I agree it's not ideal. However the alternative is to painstakingly go through every file and import the required types manually. Adding an `import * as LH from '...'` to every js file is something we could automate and do once.

The best time to not use global variables is before you start, the second best time is now (or whenever they finally fix type module imports :) The `LH` global type continues to make things more complicated than they need to be. It shouldn't take convention to prevent dependencies from stomping all over each other; imports keep it clean by default and make any conflicts local.

The `LH.` is distinctive and it wouldn't be that hard to find and replace. Biggest thing is we shouldn't have to have e.g. separate typedefs for `AuditDefn`, `Category`, `Config`, `Group`, `Json`, `Pass`, `PassJson`, and `Settings` typedefs when we could just import `Config` as a module.

**Comment by adamraine** [comment]
> Just in case we're speaking past each other, I meant as a pack/publish step. It's not useful to have them around otherwise.

Ah yup, that makes sense.

> Mixing ts and js files is pretty bad when the ts files aren't intended as source files, IMO, and makes the codebase less self evident. We're using d.ts files as the ambient declarations they already are, and it's good to have things enforced like no implementation allowed inside.

I think I would support keeping our handwritten `d.ts` more or less as they are now. Then we generate the remaining `d.ts` files next to the corresponding `.js` files like you described as part of the pack/publish step.

> The best time to not use global variables is before you start, the second best time is now (or whenever they finally fix type module imports :) The LH global type continues to make things more complicated than they need to be. It shouldn't take convention to prevent dependencies from stomping all over each other; imports keep it clean by default and make any conflicts local.
> 
> The LH. is distinctive and it wouldn't be that hard to find and replace. Biggest thing is we shouldn't have to have e.g. separate typedefs for AuditDefn, Category, Config, Group, Json, Pass, PassJson, and Settings typedefs when we could just import Config as a module.

Not sure I understand. Are you saying that `import * as LH from '...'` doesn't eliminate all complications with our current global `LH` approach? Would it be better to do individual module imports like `import Config from '../../types/config'`?

File: tsconfig-base.json [hunk 4]
```diff
@@ -8,7 +8,7 @@
     // directory structure relative to rootDir. No effect on include, exclude, etc.
     "emitDeclarationOnly": true,
     "declarationMap": true,
-    "outDir": ".tmp/tsbuildinfo/",
+    "outDir": "dist/tsbuildinfo/",
```
Comment (connorjclark) [inline tsconfig-base.json:11]: let's revert this - dist should just be build output stuff

File: .gitignore [hunk 5]
```diff
@@ -60,3 +60,14 @@ proto/scripts/*_processed.json
...
+# ignore generated d.ts files
+core/**/*.d.ts
+core/**/*.d.cts
+clients/**/*.d.ts
+cli/**/*.d.ts
+build/**/*.d.ts
+third-party/**/*.d.ts
+report/generator/**/*.d.ts
+*.d.cts
+*.d.ts
```
Comment (connorjclark) [inline .gitignore:72]: these last two would cover the rest? maybe need `**/` as prefix? can exclude `types/` folder.

File: types/externs.ts (hunk 2 above)
Comment (adamraine) [reply types/externs.ts]: #14537

File: types/lh.d.ts [hunk 6]
```diff
@@ -0,0 +1,81 @@
+/**
```
Comment (connorjclark) [inline types/lh.d.ts:1]: is `lh.d.ts` preferred over `lighthouse.d.ts`?

**Review by connorjclark — APPROVED** [review]
seems good, but Brendan should have the final call here

File: types/env/global.d.ts (hunk 1 above)
Comment (connorjclark) [reply types/env/global.d.ts:3]: is there an update on this? More PRs to come?

File: types/lh.d.ts (hunk 6 above)
Comment (adamraine) [reply types/lh.d.ts:1]: I don't have a string preference. I thought `lh.d.ts` made more sense because we refer to this stuff via the `LH` global.

File: types/env/global.d.ts (hunk 1 above)
Comment (adamraine) [reply types/env/global.d.ts:3]: The changes in this PR should provide enough typings for the normal exposed api. We should still eliminate this global though.

Doing so would be done in a follow up, and probably wouldn't be a breaking change (although it would touch many files)

File: package.json [hunk 7]
```diff
@@ -15,7 +15,7 @@
   "scripts": {
     "prepack": "yarn build-report --standalone --flow --esm",
     "build-all": "npm-run-posix-or-windows build-all:task",
-    "build-all:task": "yarn build-report && yarn build-cdt-lib && yarn build-devtools && concurrently 'yarn build-extension' 'yarn build-lr' 'yarn build-viewer' 'yarn build-treemap' 'yarn build-smokehouse-bundle' && yarn build-pack",
+    "build-all:task": "yarn build-report && yarn build-cdt-lib && yarn build-devtools && yarn build-types && concurrently 'yarn build-extension' 'yarn build-lr' 'yarn build-viewer' 'yarn build-treemap' 'yarn build-smokehouse-bundle' && yarn build-pack",
```
Comment (brendankenny) [inline package.json:18]: Maybe this is where "build all" is ambiguous, but I use it 
when I mean "build all our things", but I wouldn't want it to include all the generated type files copied alongside source files. Should it be part of `prepack` instead? I don't think there's any reason local development would ever want them in place, you really just need them there so relative paths work when Lighthouse is a dependency.

File: package.json [hunk 8]
```diff
@@ -29,9 +29,11 @@
     "build-sample-reports": "yarn build-report && node build/build-sample-reports.js",
     "build-treemap": "node ./build/build-treemap.js",
     "build-viewer": "node ./build/build-viewer.js",
+    "build-types": "tsc --build tsconfig-all.json && rsync -a .tmp/tsbuildinfo/ ./ --include='*.d.ts' --include='*.d.cts' --exclude='*.map' --exclude='*.tsbuildinfo'",
```
Comment (brendankenny) [inline package.json:32]:
```suggestion
    "build-types": "yarn type-check && rsync -a .tmp/tsbuildinfo/ ./ --include='*.d.ts' --include='*.d.cts' --exclude='*.map' --exclude='*.tsbuildinfo'",
```
?

File: package.json [hunk 9]
```diff
@@ -29,9 +29,11 @@
...
     "reset-link": "(yarn unlink || true) && yarn link && yarn link lighthouse",
     "c8": "bash core/scripts/c8.sh",
     "clean": "rm -r dist proto/scripts/*.json proto/scripts/*_pb2.* proto/scripts/*_pb.* proto/scripts/__pycache__ proto/scripts/*.pyc *.report.html *.report.dom.html *.report.json *.devtoolslog.json *.trace.json shared/localization/locales/*.ctc.json || true",
+    "clean-types": "git clean -xf '*.d.ts' '*.d.cts' -e 'node_modules/' -e 'dist/' -e '.tmp/'",
```
Comment (brendankenny) [inline package.json:36]: should this also exclude `types/`? Would be a bummer to clear out an in-progress file that's not committed yet

File: types/artifacts.d.ts [hunk 10]
```diff
@@ -12,177 +14,177 @@ import speedline from 'speedline-core';
 import TextSourceMap from '../core/lib/cdt/generated/SourceMap.js';
 import {ArbitraryEqualityMap} from '../core/lib/arbitrary-equality-map.js';
 import type { TaskNode as _TaskNode } from '../core/lib/tracehouse/main-thread-tasks.js';
-import AuditDetails from './lhr/audit-details'
-import Config from './config';
-import Gatherer from './gatherer';
-import {IcuMessage} from './lhr/i18n';
-import LHResult from './lhr/lhr'
-import Protocol from './protocol';
+import AuditDetails from './lhr/audit-details.js'
```
Comment (brendankenny) [inline types/artifacts.d.ts:17]: Are the `.js` extensions from an earlier iteration of the PR? I don't believe they're necessary and not having an extension is a nice signal that it's importing a non-source file (vs tsc's magic handling of `.js` paths leading to `.ts` files)

File: types/lhr/audit-result.d.ts [hunk 11]
```diff
@@ -4,10 +4,10 @@
  * Unless required by applicable law or agreed to in writing, software distributed under the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the License for the specific language governing permissions and limitations under the License.
  */
 
-import {FormattedIcu} from './i18n';
-import AuditDetails from './audit-details';
+import {FormattedIcu} from './i18n.js';
+import AuditDetails from './audit-details.js';
 
-interface ScoreDisplayModes {
+export interface ScoreDisplayModes {
```
Comment (brendankenny) [inline types/lhr/audit-result.d.ts:10]: Reason to export these? They don't appear to be used outside this file

File: types/env/global.d.ts [hunk 12]
```diff
@@ -0,0 +1,3 @@
+export * from '../lh.js';
```
Comment (brendankenny) [inline types/env/global.d.ts:1]: license header (and fileoverview, probably)

File: types/artifacts.d.ts [hunk 13]
```diff
@@ -12,177 +14,177 @@ import speedline from 'speedline-core';
...
-export interface Artifacts extends BaseArtifacts, GathererArtifacts {}
+interface Artifacts extends Artifacts.BaseArtifacts, Artifacts.GathererArtifacts {}
 
-export type FRArtifacts = Util.StrictOmit<Artifacts,
-  | 'Fonts'
-  | 'Manifest'
-  | 'MixedContent'
-  | keyof FRBaseArtifacts
->;
+declare module Artifacts {
```
Comment (brendankenny) [inline types/artifacts.d.ts:28]: What is the restructuring in here for? It's a lot of churn but it doesn't seem like anything fundamentally changes (just `import Artifacts` vs `import {Artifacts}`)

File: package.json (hunk 9 above)
Comment (adamraine) [reply package.json:36]: In theory it should only include files that are gitignored, but not a bad idea

File: types/artifacts.d.ts (hunk 10 above)
Comment (adamraine) [reply types/artifacts.d.ts:17]: In my testing they were necessary once the package is installed as a dependency. Not 100% sure why though.

File: types/lhr/audit-result.d.ts (hunk 11 above)
Comment (adamraine) [reply types/lhr/audit-result.d.ts:10]: Might have been from an older version of this PR, I'll try removing.

File: types/artifacts.d.ts (hunk 13 above)
Comment (adamraine) [reply types/artifacts.d.ts:28]: This was also from an earlier version, I'll try without this update

File: types/artifacts.d.ts (hunk 10 above)
Comment (adamraine) [reply types/artifacts.d.ts:17]: So I did a little further investigation. Importing types from `'./utility-types'` without the `.js` extension works fine, however importing types from `'./user-flow'` without the `.js` extension causes issues.

I think it's related to `user-flow.d.ts` importing types from an actual `.js` file.

Comment (brendankenny) [reply types/artifacts.d.ts:17]: What error are you seeing? I removed all the `.js`s from imports where there wasn't actually a `.js` file and the only errors I get when importing lighthouse (trying both default `lighthouse()` and `startFlow()`) are on global `LH` and puppeteer conflicts

Comment (adamraine) [reply types/artifacts.d.ts:17]: When importing the package into another project, we don't get any type checking info.

Basically any type imported from a file not using the .js extension is treated as `any`

Comment (brendankenny) [reply types/artifacts.d.ts:17]: example lighthouse code showing a typescript error on a function parameter

working for me, and this goes
`import {startFlow} from 'lighthouse'`
`index.d.ts` -> `import * as LH from "../types/lh.js"` to get `LH.UserFlow.Options`
`lh.d.ts` -> `import UserFlow_ from './user-flow'` to get `UserFlow`
`user-flow.d.ts` -> `import {Flags} from './externs'` to get the `Flags` type

so two hops through imports with no file extension, and `user-flow.d.ts` has imports with and without an extension, because it needs the actual js `'../core/user-flow.js'` file.

Comment (adamraine) [reply types/artifacts.d.ts:17]: I think I figured out what's going on...

I was doing my testing in a project with a `tsconfig.json` containing `"moduleResolution": "NodeNext"`. If I change it to `"moduleResolution": "node"` my results match what you are seeing.

Seems like we need the `.js` extension on our module imports so type checking works in ts projects set with `"moduleResolution": "NodeNext"`.

Comment (connorjclark) [reply types/artifacts.d.ts:17]: > Seems like we need the .js extension on our module imports so type checking works in ts projects set with "moduleResolution": "NodeNext".

That's the purpose of nodenext - to opt into standards compliant ES modules. No need to guess the extension, and TS prefers to not modify authored code when possible so they chose to require .js and skip any rewriting step based on the output format.

Comment (adamraine) [reply types/artifacts.d.ts:17]: @brendankenny WDYT about keeping the `.js` extensions for compatibility with `NodeNext`?

Comment (benschwarz) [reply types/artifacts.d.ts:17]: Seems good from a compatibility perspective to keep the `.js` extensions.

**Review by connorjclark — APPROVED** [review]
Didn't come across any problems when using the types generated by this PR. Nice work! This should be so much better for our Node users.

File: .github/workflows/ci.yml [hunk 14]
```diff
@@ -31,6 +31,7 @@ jobs:
         node-version: 16.x
 
     - run: yarn install --frozen-lockfile --network-timeout 1000000
+    - run: yarn type-check
```
[the remark this answers is not in the document]
Comment (adamraine) [reply .github/workflows/ci.yml:34]: `yarn type-check` is run as part of `yarn build-all` now, that would have made completely `yarn type-check` redundant and it would be confusing to receive type check errors as part of the build step.

**Comment by adamraine** [comment]
> Maybe add a note to RunnerResult artifacts: Artifacts; that a subset of artifacts are considered unstable? We could mark those as @deprecated to get IDEs to call it out, but that seems like it could have undesirable effects with certain tooling.

If we don't want people to use them, maybe `@internal` could work? I'm good with just a comment though.

File: types/artifacts.d.ts (hunk 10 above)
Comment (brendankenny) [reply types/artifacts.d.ts:17]: Ugh, I guess there's not any way around this.

Just to be clear about my objection:
- Unlike the usual `.ts` -> `.js` import, there is no js file to import, and there won't be after a build step, because these are imports of `d.ts` files purely for types. Even if we were doing a pure-ts compile, `d.ts` files never get compiled or emitted.
- there's no issue of the compiler not re-writing anything because these imports are _also_ in `d.ts` files, which again are never compiled or emitted, so there's no output to rewrite the paths to

As far as `tsc` is concerned, these files only exist at type-checking time and then cease to exist, and they are definitely not es modules.

`nodenext` still has a ton of issues (a number of fixes queued up before 5.0), but as far as I can tell they aren't going to budge on this unless a more significant contingent of jsdoc users object. I guess we'll have to live with these js imports.

File: flow-report/tsconfig.json [hunk 15]
```diff
@@ -19,7 +19,7 @@
     "**/*.ts",
     "**/*.tsx",
     "./types",
-    "../types/test.d.ts",
+    "../types/env/test.d.ts",
```
Comment (brendankenny) [inline flow-report/tsconfig.json:22]: maybe `types/dev/*.d.ts`? Or `types/internal/*.d.ts`? I get that idea of `env`, but it's not the first thing I think of (there's no reason a child project can't inherit an `env`), so the others may be more descriptive

Comment (adamraine) [reply flow-report/tsconfig.json:22]: `types/internal/*.d.ts` sounds good to me

**Comment by brendankenny** [comment]
> Didn't come across any problems when using the types generated by this PR. Nice work! This should be so much better for our Node users.

We're not quite there even for a basic user flow. Sprinkling in some more `LH` imports will get things working for most under `"moduleResolution": "node"`, but if everyone is really as eager to use `"moduleResolution": "NodeNext"`  as claim they are :P, there's still a couple of tricky parts to fix before it can be correctly imported

**Comment by connorjclark** [comment]
> > Maybe add a note to RunnerResult artifacts: Artifacts; that a subset of artifacts are considered unstable? We could mark those as @deprecated to get IDEs to call it out, but that seems like it could have undesirable effects with certain tooling.
> 
> If we don't want people to use them, maybe `@internal` could work? I'm good with just a comment though.

"@internal" is the right semantics. It'll at least show up as the first line of the jsdoc hover tooltip. A comment on the artifacts field is also good I think, let's do both?


> We're not quite there even for a basic user flow. Sprinkling in some more LH imports will get things working for most under "moduleResolution": "node", but if everyone is really as eager to use "moduleResolution": "NodeNext" as claim they are :P, there's still a couple of tricky parts to fix before it can be correctly imported

I only verified IDE support looked good... which it does for user flows AFAICT. but yeah using tsc with nodenext and node16 is giving me various issues about file extensions missing:

```
Relative import paths need explicit file extensions in EcmaScript imports when '--moduleResolution' is 'node16' or 'nodenext'. Did you mean '../core/gather/fetcher.js'?
```

File: types/externs.d.ts [hunk 16]
```diff
@@ -71,7 +71,10 @@ export interface CliFlags extends Flags {
 export interface RunnerResult {
   lhr: LHResult;
   report: string|string[];
-  /** WARNING: artifacts are not guaranteed to be stable. The structure is subject to change in minor releases. */
+  /**
+   * @internal
+   * WARNING: artifacts are not guaranteed to be stable. The structure is subject to change in minor releases.
```
Comment (connorjclark) [inline types/externs.d.ts:76]: woah let's not suggest there are _all_ unstable. We have a notion of public/private artifacts after all

**Comment by adamraine** [comment]
(e2e test failure is unrelated)

The errors @connorjclark noticed were valid and have been fixed in https://github.com/GoogleChrome/lighthouse/pull/14441/commits/ba3104f26763e5de0c0a232115043e413f814e60. I also added a recipe doc + test for a project using typescript and nodenext module resolution https://github.com/GoogleChrome/lighthouse/pull/14441/commits/324012a4c01e966c145f73607764d972286d509e.

At this point, I think this PR is ready to merge.

File: docs/recipes/type-checking/tsconfig.json [hunk 17]
```diff
@@ -0,0 +1,12 @@
+{
+  "compilerOptions": {
+    "noEmit": true,
+    "target": "esnext",
+    "module": "esnext",
```
Comment (brendankenny) [inline docs/recipes/type-checking/tsconfig.json:5]: I don't believe it technically affects anything because there's no emit, but `module` and `moduleResolution` should both be `'NodeNext'` if one of them is (I think that might be an error in the future?)

File: docs/recipes/type-checking/use-types.ts [hunk 18]
```diff
@@ -0,0 +1,22 @@
+import * as lhApi from 'lighthouse';
```
Comment (brendankenny) [inline docs/recipes/type-checking/use-types.ts:1]: can you add license and a short readme to the directory? It is supposed to be a recipe, not just a test :)

File: docs/recipes/type-checking/use-types.ts [hunk 19]
```diff
@@ -0,0 +1,22 @@
...
+
+const flow: lhApi.UserFlow = await lhApi.startFlow(page, {config});
+
+await flow.navigate('https://example.com');
+
+await flow.startTimespan({name: 'Click button'});
+await page.click('button');
+await flow.endTimespan();
+
+await flow.snapshot({name: 'New page state'});
```
Comment (brendankenny) [inline docs/recipes/type-checking/use-types.ts:22]: generate the report as well?

**Review by brendankenny — APPROVED** [review]
Very exciting! 🎉 

The test is a great addition. It's possible it will end up annoying detecting (additional) type errors in this roundabout way (does anyone run the docs tests except in CI?) but let's see if it's actually a problem and if there's more we can do to mitigate it (e.g. lint import paths in `d.ts` files or whatever)
